sdks/java/core/src/main/java/org/apache/beam/sdk/util/state/StateSpec.java

```diff
@@ -36,4 +37,10 @@
    * Use the {@code binder} to create an instance of {@code StateT} appropriate for this address.
    */
   StateT bind(String id, StateBinder<? extends K> binder);
+
+  /**
+   * Given {code coders} are inferred from type arguments defined for this class.
+   * @param coders Array of coders indexed by the type arguments order.
+   */
+  void offerCoders(Coder[] coders);
```

Comment:
This is nice, I think. I want to spend a minute pondering it.

What is the contract if it is called more than once? What if some of the coders cannot be provided but other can? These are issues that are very similar to `TypedPValue.finishSpecifying()` and the coder registry's internal logics.